### PR TITLE
Navbar custom items

### PR DESCRIPTION
**Pull request by afiq0209**
**Changes**
Add custom component to navbar items

- [delete] branch

**Changelog entry**

-  Description of change in `Navbar`
- Allowing custom component to be added as navbar items.

**Additional information**

- You may refer to this [BOOKINGSG-4513](https://jira.ship.gov.sg/browse/BOOKINGSG-4513)
- Figma reference: [link](https://www.figma.com/file/UUg9P4nOScnJXfrA2By2aw/Facilities-%E2%80%94-Admin-Portal?type=design&node-id=6178-248501&mode=design&t=nhfcQgLYnfO0C9FB-0)

### CODE REVIEW

File: stories/navbar/navbar.stories.mdx [hunk 1]
```diff
@@ -264,6 +267,128 @@ You can also choose to render a custom component of your choice as an action but
...
+                                                    marginBottom: "1rem",
+                                                }}
+                                                onClick={() => {
+                                                    navbarRef.current.dismissDrawer();
+                                                }}
+                                            >
+                                                Click to close the drawer
+                                            </Button.Small>
+                                        </CustomComponentWrapper>
+                                        <hr />
```
Comment (qroll) [inline stories/navbar/navbar.stories.mdx:339]: can trouble you to extract this into a component in `doc-elements.tsx` to keep the story markup simpler

File: stories/navbar/navbar.stories.mdx [hunk 2]
```diff
@@ -264,6 +267,128 @@ You can also choose to render a custom component of your choice as an action but
     </Story>
 </Canvas>
 
+<Heading3>Custom component items</Heading3>
+
+You can also choose to render a custom component of your choice as an items.
```
Comment (qroll) [inline stories/navbar/navbar.stories.mdx:272]:
```suggestion
You can also choose to render a custom component of your choice as a navigation bar item.
```

File: stories/navbar/navbar.stories.mdx [hunk 3]
```diff
@@ -264,6 +267,128 @@ You can also choose to render a custom component of your choice as an action but
...
+<Canvas>
+    <Story name="Custom items">
+        {() => {
+            const navbarRef = useRef();
+            const [selected, setSelected] = useState("home");
+            return (
+                <Navbar
+                    ref={navbarRef}
+                    items={{
+                        desktop: [
```
Comment (qroll) [inline stories/navbar/navbar.stories.mdx:283]: can you also provide an example of a custom component on desktop? doesn't have to be as complicated as the mobile one, a simple icon will do

File: src/navbar/navbar-items.tsx [hunk 4]
```diff
@@ -97,56 +97,67 @@ export const NavbarItems = <T,>({
     // =============================================================================
     const renderItems = () => {
         return items.map((item, index) => {
-            const selected = checkSelected(item);
-            const { children, options, ...otherItemAttrs } = item;
+            switch (item.itemType) {
+                case "component": {
+                    const component = (item && item.children) || null;
+                    return <div key={index}>{component}</div>;
```
Comment (qroll) [inline src/navbar/navbar-items.tsx:103]: this should be a list item as it's part of a list

```suggestion
                    return <li key={index}>{component}</li>;
```

**Review by qroll — CHANGES_REQUESTED** [review]

File: src/navbar/navbar.tsx [hunk 5]
```diff
@@ -52,9 +59,21 @@ const Component = <T,>(
     const [showDrawer, setShowDrawer] = useState<boolean>(false);
     const [showOverlay, setShowOverlay] = useState<boolean>(false);
     const isStretch = layout === "stretch";
+    const elementRef = useRef<HTMLDivElement>();
 
     const { primary = DEFAULT_RESOURCES.primary, secondary } = resources;
 
+    useImperativeHandle(
+        ref,
+        () =>
```
Comment (keithtxw) [inline src/navbar/navbar.tsx:68]:
```suggestion
        () => ({
            ...elementRef.current,
            dismissDrawer: () => {
                dismissDrawer();
            }
        })
```

File: src/navbar/types.ts [hunk 6]
```diff
@@ -4,10 +4,19 @@ import { TextLinkProps } from "../text";
...
-export interface NavItemProps<T> extends NavItemCommonProps<T> {
+
+export interface NavItemLinkProps<T> extends NavItemCommonProps<T> {
+    itemType?: "link";
     subMenu?: NavItemCommonProps<T>[] | undefined;
 }
 
+export interface NavItemComponentProps {
+    itemType: "component";
+    children: JSX.Element;
```
Comment (keithtxw) [inline src/navbar/types.ts:15]: actually `children` in the `TextLinkProps` allows for JSX elements too. I would go with using `NavItemLinkProps` as a single interface with an optional prop called `renderAs` where it only allows `"link" | "custom-component"` and it defaults to `"link"`

**Review by keithtxw — CHANGES_REQUESTED** [review]

File: src/navbar/navbar.tsx (hunk 5 above)
Comment (qroll) [reply src/navbar/navbar.tsx]: actually using `Object.assign` is required, the spread syntax loses the div element attributes. this issue was encountered with `Accordion.Item`

**Review by qroll — APPROVED** [review]

**Review by keithtxw — APPROVED** [review]

**Pull request merged**